src/tools/compiletest/src/runtest.rs

```diff
@@ -1999,10 +2011,14 @@ impl<'test> TestCx<'test> {
         }
 
         if !is_rustdoc {
-            if self.config.target == "wasm32-unknown-unknown" {
+            if self.config.target == "wasm32-unknown-unknown"
+            || !self.use_dynamic_linking() {
                 // rustc.arg("-g"); // get any backtrace at all on errors
             } else if !self.props.no_prefer_dynamic {
                 rustc.args(&["-C", "prefer-dynamic"]);
+                if self.config.target.contains("vxworks") {
```

Comment:
I'm ok with the other changes, but this logic should be in the compiler, not the test code. I.e. `-C prefer-dynamic` on VxWorks should imply `-C target-feature=-crt-static`. Otherwise the tests may pass but "real world" use will lead to errors.

Reply:
@BaoshanPang 

> You wrote:
> The reason that we need to change the code in test framework is that in testing the auxiliary library is built in shared library only  which makes some “magic” happens inside rust,
> 
> It is not about if you are using option “-C prefer-dynamic” or not, what I noticed is if the auxiliary library is built in shared library then the shared library of “std” is linked into executable,
> 
> and if the auxiliary library is built in static library then the static library of “std” is linked into executable.

OK, then I'm ok with this change.